api/middleware.go

```diff
@@ -0,0 +1,245 @@
+	// set it to the default value.
+	v := reflect.ValueOf(p).Elem()
+	for i := 0; i < v.NumField(); i++ {
+		f := v.Field(i)
+		if f.Kind() == reflect.Ptr && f.IsNil() {
+			switch v.Type().Field(i).Name {
+			case "Limit":
+				f.Set(reflect.ValueOf(&defaultLimit))
+			case "Offset":
+				f.Set(reflect.ValueOf(&defaultOffset))
+			case "BucketSizeSeconds":
+				f.Set(reflect.ValueOf(&defaultBucketSizeSeconds))
+			}
+		}
+	}
+
+	// Iterate through the struct fields. If the field name equals "Limit" and it's of the right type, clamp it.
+	for i := 0; i < v.NumField(); i++ {
+		f := v.Field(i)
+		if f.Kind() == reflect.Ptr && !f.IsNil() {
+			switch v.Type().Field(i).Name { //nolint:gocritic // allow single-case switch for future expansions
+			case "Limit":
+				if v.Type().Field(i).Type == reflect.TypeOf(&maxLimit) && *f.Interface().(*uint64) > maxLimit {
+					*f.Interface().(*uint64) = maxLimit
+				}
+			}
+		}
+	}
+}
+
+// FixDefaultsAndLimits modifies pagination parameters of the request in-place:
+// If they're missing, it assigns them default values, and if they exist, it
+// clamps them within the allowed limits.
+// Both of these should be done by oapi-codegen, but it doesn't do it yet.
+//
+// _operationID is unused, but is required to match the StrictHandlerFunc signature.
+// It takes values like "GetConsensusTransactions".
+func FixDefaultsAndLimitsMiddleware(next apiTypes.StrictHandlerFunc, _operationID string) apiTypes.StrictHandlerFunc {
+	return func(ctx context.Context, w http.ResponseWriter, r *http.Request, args interface{}) (interface{}, error) {
+		// Create a new struct of the same type as `args` and copy the values from `args` into it.
+		// This new struct will be "addressable" (= modifiable), unlike `args`.
+		argsV := reflect.ValueOf(args)
+		v := reflect.New(argsV.Type())
+		v.Elem().Set(argsV)
+
+		// Use reflection to check if `args` has a Params field.
+		// If it does, we'll use it to set default values and limits.
+		if v.Kind() == reflect.Ptr || v.Kind() == reflect.Interface {
+			v = v.Elem()
+		}
+		if v.Kind() == reflect.Struct {
+			f := v.FieldByName("Params")
+			if f.IsValid() {
+				fixDefaultsAndLimits(f.Addr().Interface())
+			}
+		}
+
+		// Call the next middleware in the chain with fixed `args`.
+		return next(ctx, w, r, v.Interface())
+	}
+}
+
+// Find the json annotation on a struct field, and return the json specified
+// name if available, otherwise, just the field name.
+func jsonName(f reflect.StructField) string {
+	fieldName := f.Name
+	tag := f.Tag.Get("json")
+	if tag != "" {
+		tagParts := strings.Split(tag, ",")
+		name := tagParts[0]
+		if name != "" {
+			fieldName = name
+		}
+	}
+	return fieldName
+}
+
+// ParseBigIntParamsMiddleware fixes the parsing of request parameters of type *BigInt.
+// oapi-codegen does not really support reading inputs into structs (except Date and Time, two hardcoded exceptions).
```

Comment:
Does openapi validate other param types like `staking.Address` and `signature.PublicKey`?

Reply:
Surprisingly, yes. Here's my belief as to why; now added into comments:
```
// NOTE: oapi-codegen _does_ support some custom type parsing, so we don't need to patch their parsing here.
// Date and Time are two hardcoded supported structs. Also, non-struct typedefs (like `type Address [21]byte`,
// which is our `staking.Address`) work fine.
```

When I was debugging oapi-codegen, structs are handled on a special and convoluted codepath, so I'm 90% confident in this explanation. I haven't verified with a debugger though.